File: py/knn.py
from collections import defaultdict
import heapq
import math
from functools import reduce

# [claude review] zip silently truncates if lists are different lengths.
# Fine if you trust your data, but worth knowing.
def euclidean_distance(
    x: list[float],
    y: list[float]
) -> float:
    return math.sqrt(sum((a - b) ** 2 for a, b in zip(x, y)))

# [claude review] same zip truncation note as above.
def manhattan_distance(
    x: list[float],
    y: list[float]
) -> float:
    return sum(abs(a - b) for a, b in zip(x, y))

# returns indexes of k nearest
#
# Uses a max-heap of size k. We push (-dist, idx) so the farthest
# candidate is always at the top. When a closer point is found,
# pop the farthest and push the new one. O(n log k) total.
def find_k_nearest(
    query: list[float],
    training_data: list[tuple[list[float], str]],
    k: int,
) -> list[int]:
    heap: list[tuple[float, int]] = []  # max-heap via negated distances
    for i, (point, _label) in enumerate(training_data):
        dist = euclidean_distance(point, query)
        if len(heap) < k:
            heapq.heappush(heap, (-dist, i))
        elif dist < -heap[0][0]:
            heapq.heapreplace(heap, (-dist, i))

    return [i for _, i in sorted(heap, reverse=True)]

# [claude review] Works but reduce with a mutating accumulator is
# unidiomatic Python. A loop or Counter is more natural:
#
#   from collections import Counter
#   counts = Counter(training_data[i][1] for i in idxs)
#   return counts.most_common(1)[0][0]
def classify(
    idxs: list[int],
    training_data: list[tuple[list[float], str]],
) -> str:
    def count_labels(acc: defaultdict[str, int], idx: int):
        acc[training_data[idx][1]] += 1
        return acc

    freq_map : dict[str, int] = reduce(count_labels, idxs, defaultdict(int))
    max_idx = max(freq_map, key = lambda label: freq_map[label])
    return max_idx


def predict(
    query: list[float],
    training_data: list[tuple[list[float], str]],
    k: int,
) -> str:
    k_nearest_points = find_k_nearest(query, training_data, k)
    return classify(k_nearest_points, training_data)

def evaluate(
    test_data: list[tuple[list[float], str]],
    training_data: list[tuple[list[float], str]],
    k: int,
) -> float:
    correct = sum(1 for features, label in test_data if predict(features, training_data, k) == label)
    return correct / len(test_data)

File: cpp/src/knn.cpp
#include "star_nn/knn.h"

#include <cmath>
#include <cstddef>
#include <ranges>
#include <unordered_map>
#include <utility>

namespace star_nn {

double euclidean_distance(const Vector &a, const Vector &b) {
  double d = 0.0;
  for (auto [x, y] : std::views::zip(a, b)) {
    d += std::pow(x - y, 2);
  }
  return std::sqrt(d);
}

double manhattan_distance(const Vector &a, const Vector &b) {
  double d = 0.0;
  for (auto [x, y] : std::views::zip(a, b)) {
    d += std::abs(x - y);
  }
  return d;
}

std::vector<std::size_t> find_k_nearest(const Vector &query,
                                        const Dataset &training_data,
                                        std::size_t k) {

  std::vector<std::pair<double, size_t>> candidates;
  for (std::size_t i = 0; i < training_data.size(); ++i) {
    const auto &dp = training_data[i];

    auto dist = euclidean_distance(dp.features, query);

    if (candidates.size() < k || candidates.back().first > dist) {
      auto item = std::make_pair(dist, i);

      auto pos = std::lower_bound(
          candidates.begin(), candidates.end(), item,
          [](const auto &a, const auto &b) { return a.first < b.first; });
      candidates.insert(pos, item);

      if (candidates.size() > k) {
        candidates.pop_back(); // remove farthest
      }
      continue;
    }
  }

  // return candidates
  std::vector<size_t> output(k);
  for (std::size_t i = 0; i < candidates.size(); ++i) {
    output[i] = candidates[i].second;
  }
  return output;
}

Label classify(const std::vector<std::size_t> &neighbor_indices,
               const Dataset &training_data) {
  std::unordered_map<Label, int> counts;

  int max_count = -1;
  Label max_label = "";

  for (const auto i : neighbor_indices) {
    const auto label = training_data[i].label;

    counts[label] += 1;
    if (counts[label] > max_count) {
      max_count = counts[label];
      max_label = label;
    }
  }

  return max_label;
}

Label predict(const Vector &query, const Dataset &training_data,
              std::size_t k) {
  const auto k_nearest = find_k_nearest(query, training_data, k);
  const auto label = classify(k_nearest, training_data);

  return label;
}

double evaluate(const Dataset &test_data, const Dataset &training_data,
                std::size_t k) {
  std::size_t correct = 0;

  for (const auto &dp : test_data) {
    if (predict(dp.features, training_data, k) == dp.label) {
      ++correct;
    }
  }

  return static_cast<double>(correct) / test_data.size();
}

} // namespace star_nn
